File: android/src/main/kotlin/com/sabiantech/sabian_native_common/utilities/permissions/PermissionRationale.kt
package com.sabiantech.sabian_native_common.utilities.permissions

class PermissionRationale(
        var permissions: Array<String>,
        var allMandatory: Boolean = true,
        var description: String? = null,
        var onRequested: ((PermissionRationale) -> Unit)? = null,
        var accepted: List<String>? = null,
        var denied: List<String>? = null,
        var onError: ((Exception) -> Unit)? = null,
        var areAllDenied: Boolean? = null,
        var isAnyDenied: Boolean? = null,
        var isAnyDeniedPermanently: Boolean? = null
) {
    companion object {
        fun shallowCopy(rationale: PermissionRationale): PermissionRationale {
            return PermissionRationale(rationale.permissions, rationale.allMandatory)
        }
    }


    val isEmpty: Boolean
        get() = permissions.isEmpty()

    val areAllPermissionsDenied: Boolean
        get() {
            if (areAllDenied != null)
                return areAllDenied!!
            val denied = denied ?: listOf()
            return permissions.size == denied.size
        }


    val isAnyPermissionDenied: Boolean
        get() {
            if (isAnyDenied != null)
                return isAnyDenied!!
            val denied = denied ?: listOf()
            return permissions.any {
                denied.contains(it)
            }
        }


    val isAnyPermissionAccepted: Boolean
        get() {
            if (areAllPermissionsDenied)
                return false
            val accepted = accepted ?: listOf()
            return permissions.any {
                accepted.contains(it)
            }
        }

    fun isPermissionDenied(permission: String): Boolean {
        return denied?.contains(permission) ?: false
    }

    fun isPermissionAccepted(permission: String): Boolean {
        return accepted?.contains(permission) ?: false
    }
}

File: ios/Classes/operations/permissions/PermissionRationale.swift
//
//  PermissionRationale.swift
//  Runner
//
//  Created by Sweet Pea on 20/11/2023.
//

import Foundation
class PermissionRationale {
       var permissions: [SabianPermissionsType] = []
       var allMandatory: Bool = true
       var description: String? = nil
       var onRequested: ((PermissionRationale) -> Void)? = nil
       var accepted: [SabianPermissionsType]? = nil
       var denied: [SabianPermissionsType]? = nil
       var onError: ((Error) -> Void)? = nil
       var areAllDenied: Bool? = nil
       var isAnyDenied: Bool? = nil
       var isAnyDeniedPermanently: Bool? = nil
    
    init(_ permissions : [SabianPermissionsType]){
        self.permissions = permissions
    }
    
    var areAllPermissionsDenied : Bool{
        get {
            if(areAllDenied != nil){
                return areAllDenied!
            }
            let denied = self.denied ?? []
            return permissions.count == denied.count
        }
    }
    
    var isAnyPermissionDenied : Bool {
        get {
            if(isAnyDenied != nil){
                return isAnyDenied!
            }
            let denied = denied ?? []
            return permissions.contains(where: { per in
                    denied.contains(per)
                })
        }
    }
    
    var isAnyPermissionAccepted : Bool {
        get {
            if(areAllPermissionsDenied){
                return false
            }
            let accepted = accepted ?? []
            return permissions.contains(where: { per in
                    accepted.contains(per)
                })
        }
    }

    var areAllPermissionsAccepted : Bool {
        get {
            if(isAnyPermissionDenied){
                return false
            }
            return accepted?.count == permissions.count
        }
    }


      func isPermissionDenied(permission: SabianPermissionsType) -> Bool {
          return denied?.contains(permission) ?? false
      }

      func isPermissionAccepted(permission: SabianPermissionsType) -> Bool {
          return accepted?.contains(permission) ?? false
      }
}
